src/crystal/system/unix/evented/waiters.cr

```diff
@@ -4,19 +4,31 @@
 # Thread safe: mutations are protected with a lock, and race conditions are
 # handled through the ready atomic.
 struct Crystal::Evented::Waiters
-  @ready = Atomic(Bool).new(false)
+  {% if flag?(:preview_mt) %}
+    @ready = false
+    @closed = false
+  {% end %}
   @lock = SpinLock.new
   @list = PointerLinkedList(Event).new
 
+  # Adds an event to the waiting list. May return false immediately if another
+  # thread marked the list as ready in parallel, returns true otherwise.
   def add(event : Pointer(Event)) : Bool
     {% if flag?(:preview_mt) %}
-      # check for readiness since another thread running the evloop might be
-      # trying to dequeue an event while we're waiting on the lock (failure to
-      # notice notice the IO is ready)
-      return false if ready?
-
       @lock.sync do
-        return false if ready?
+        if @closed
+          # another thread closed the fd or we received a fd error or hup event:
+          # the fd will never block again
+          return false
+        end
+
+        if @ready
+          # another thread readied the fd before the current thread got to add
+          # the event: don't block and resets @ready for the next loop
+          @ready = false
+          return false
```

Comment:
**question:** If we reset `@ready`, what happens if there's another (or more) thread in the same situation, trying to add an event (while another thread holding the lock originally set `@ready = true`) 🤔

Reply:
This swap replicates the `#ready_one` behavior: it activates one waiter while the others will have to wait until the `fd` blocks, which will activate one waiter when the `fd` is ready again. This is to avoid a thundering herd issue —the libevent evloop also only resumes a single waiter (i.e. `IO::Evented#resume_read`).

Still, there might be an edge case (applies to `IO::Evented` too) when multiple fibers try to read from or write to the same IO object at the same time (i.e. open brawl) and the resumed fiber doesn't exhaust the kernel buffer and goes on to do something else. In that scenario the other fibers will stay blocked until a new read/write would block or an error/close event happens, despite the `fd` being ready.